pkg/executor/aggregate/agg_hash_final_worker.go

```diff
@@ -174,19 +223,75 @@ func (w *HashAggFinalWorker) receiveFinalResultHolder() (*chunk.Chunk, bool) {
 	}
 }
 
+func (w *HashAggFinalWorker) mergeResultsAndSend(ctx sessionctx.Context) error {
+	err := w.consumeIntermData(ctx)
+	if err != nil {
+		w.outputCh <- &AfFinalResult{err: err}
+		return err
+	}
+
+	w.loadFinalResult(ctx)
+	return nil
+}
+
 func (w *HashAggFinalWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup) {
 	start := time.Now()
-	defer func() {
-		if r := recover(); r != nil {
-			recoveryHashAgg(w.outputCh, r)
+	defer w.cleanup(start, waitGroup)
+
+	failpoint.Inject("enableAggSpillIntest", func(val failpoint.Value) {
+		if val.(bool) {
+			intestBeforeFinalWorkerStart()
 		}
-		if w.stats != nil {
-			w.stats.WorkerTime += int64(time.Since(start))
+	})
+
+	w.isSpilledTriggered = w.spillHelper.isSpillTriggered()
+	if w.isSpilledTriggered {
+		if w.spillHelper.checkError() {
+			return
+		}
+
+		w.loadFinalResult(ctx)
+	} else {
+		err := w.mergeResultsAndSend(ctx)
```

Comment:
remove the definition of mergeResultsAndSend seems to be more readable
``` go
	if w.isSpilledTriggered {
		if w.spillHelper.checkError() {
			return
		}
	} else {
		err := w.consumeIntermData(ctx)
		if err != nil {
			w.outputCh <- &AfFinalResult{err: err}
			return 
		}
	}
	w.loadFinalResult(ctx)
```

Reply:
> remove the definition of mergeResultsAndSend seems to be more readable
> 
> ```go
> 	if w.isSpilledTriggered {
> 		if w.spillHelper.checkError() {
> 			return
> 		}
> 	} else {
> 		err := w.consumeIntermData(ctx)
> 		if err != nil {
> 			w.outputCh <- &AfFinalResult{err: err}
> 			return 
> 		}
> 	}
> 	w.loadFinalResult(ctx)
> ```

done